onnxruntime/core/providers/webgpu/tensor/concat.cc

```diff
@@ -0,0 +1,144 @@
+#include "core/providers/webgpu/shader_variable.h"
+#include "core/providers/webgpu/webgpu_kernel.h"
+#include "core/providers/webgpu/webgpu_supported_types.h"
+
+namespace onnxruntime {
+namespace webgpu {
+
+#define WEBGPU_CONCAT_VERSIONED_KERNEL(start, end)            \
+  ONNX_OPERATOR_VERSIONED_KERNEL_EX(                          \
+      Concat,                                                 \
+      kOnnxDomain,                                            \
+      start,                                                  \
+      end,                                                    \
+      kWebGpuExecutionProvider,                               \
+      (*KernelDefBuilder::Create())                           \
+          .TypeConstraint("T", WebGpuSupportedNumberTypes()), \
+      Concat);
+
+#define WEBGPU_CONCAT_KERNEL(version)                         \
+  ONNX_OPERATOR_KERNEL_EX(                                    \
+      Concat,                                                 \
+      kOnnxDomain,                                            \
+      version,                                                \
+      kWebGpuExecutionProvider,                               \
+      (*KernelDefBuilder::Create())                           \
+          .TypeConstraint("T", WebGpuSupportedNumberTypes()), \
+      Concat);
+
+WEBGPU_CONCAT_VERSIONED_KERNEL(1, 3)
+WEBGPU_CONCAT_VERSIONED_KERNEL(4, 10)
+WEBGPU_CONCAT_VERSIONED_KERNEL(11, 12)
+WEBGPU_CONCAT_KERNEL(13)
+
+void AppendCalCulateInputIndexFunction(std::ostream& os, size_t input_count) {
+  os << "fn calculate_input_index(index: u32) -> u32 {\n"
+     << "  for (var i = 0u; i < " << input_count << "; i = i + 1u) {\n"
+     << "    if (index < " << GetElementAt("uniforms.size_in_concat_axis", "i", input_count) << ") {\n"
+     << "      return i;\n"
+     << "    }\n"
+     << "  }\n"
+     << "  return " << input_count << ";\n"
+     << "}\n";
+}
+
+void AppendAssignOutputDataFunction(std::ostream& os, gsl::span<const ShaderVariableHelper*> inputs, const ShaderVariableHelper& output) {
+  os << "fn assign_output_data(global_idx: u32, input_index: u32, indices: output_indices_t) {\n";
+  for (size_t i = 0; i < inputs.size(); ++i) {
+    if (i == 0) {
+      os << "  if (input_index == 0u) {\n";
+    } else if (i == inputs.size() - 1) {
+      os << "  } else {\n";
+    } else {
+      os << "  } else if (input_index == " << i << "u) {\n";
+    }
+    os << "     " << output.SetByOffset("global_idx", inputs[i]->GetByIndices("indices")) << ";\n";
+  }
+  os << "  }\n"
+        "}\n";
+}
+
+Status ConcatProgram::GenerateShaderCode(ShaderHelper& shader) const {
+  size_t input_count = Inputs().size();
+  std::vector<const ShaderVariableHelper*> inputs;
+  inputs.reserve(input_count);
+  for (size_t i = 0; i < input_count; ++i) {
+    inputs.push_back(&shader.AddInput("input_" + std::to_string(i), ShaderUsage::UseUniform | ShaderUsage::UseIndicesTypeAlias));
```

Comment:
>shader.AddInput("input_" + std::to_string(i), ShaderUsage::UseUniform | ShaderUsage::UseIndicesTypeAlias)); [](http://example.com/codeflow?start=22&length=107)

Why do you take the address on the returned value? ShaderHelper uses a vector to hold such vars. Then address of the vars can get changed when the vector is resizing.

Reply:
A `ShaderVariableHelper` object is stateful and non-copyable. `shader.AddInput()` returns a reference to `ShaderVariableHelper` .

do you suggest to use `std::vector<std::reference_wrapper<ShaderVariableHelper>>` instead of `std::vector<const ShaderVariableHelper*>`?

Reply:
I think this is a bug now, because the `const ShaderVariableHelper*` points could suddenly become invalid.

Reply:
the lifecycle/ownership is as:

every instance of ShaderVariableHelper is created by and owned by ShaderHelper, and the ShaderHelper is available in the function call of GenerateShaderCode().

ShaderHelper::AddInput() returns a reference to an instance of ShaderIndicesHelper, which is stored in a std::vector<std::unique_ptr<ShaderIndicesHelper>> as a member in ShaderHelper.

"The pointer `const ShaderVariableHelper*` becomes invalid" only happens if `GenerateShaderCode()` stores the pointer and use it later outside of this function.